### ISSUE TITLE
Bust cache after a release?

### ISSUE THREAD

**Issue by scambier** [antfu/vite-plugin-pwa#33]
Hello, I'm not well versed in PWA settings, and your plugin works really well as a 0-config tool (thanks for that).

However, I'm faced with a cache issue when I'm redeploying my app. In that case, I had to rename several api endpoints on my backend, but once the Vue app was built and deployed, all I got was 404 errors because I was still served the app from the cache. Since it's still in dev I simply manually unregistered the worker, but I what to do if that happens in production?

If I understand correctly, even if all my .js files and assets have a versioned hash in their name, as long as the index.html file is cached, it will still serve the old assets.

Is there a configuration I can setup (maybe it's still WIP on your side) to fix this, or if applicable, is there something you recommend? I found several solutions for this issue, but nothing looks really clean or fail-proof.

**Comment by userquin** [comment]
Hi,

There is a pending think to do about refresh on new content detected.

You can use [register-service-worker](https://github.com/yyx990803/register-service-worker), and do a manual installation of service worker via `VitePWA` configuration on `vite.config.ts` file, something like this:

 `useServiceWorker.ts`
```typescript
import type { Ref } from 'vue'

import { register } from 'register-service-worker'
import { ref } from 'vue'

// https://medium.com/google-developer-experts/workbox-4-implementing-refresh-to-update-version-flow-using-the-workbox-window-module-41284967e79c
// https://github.com/yyx990803/register-service-worker
// https://developers.google.com/web/fundamentals/primers/service-workers/high-performance-loading
export const useServiceWorker = (): {
  appNeedsRefresh: Ref<boolean>
  offlineAppReady: Ref<boolean>
} => {
  const offlineAppReady = ref(false)
  const appNeedsRefresh = ref(false)

  register(
    '/sw.js',
    {
      registrationOptions: { scope: '/' },
      updated: async() => { appNeedsRefresh.value = true },
      cached: () => { offlineAppReady.value = true },
    },
  )

  return {
    appNeedsRefresh,
    offlineAppReady,
  }
}
```

then on your`main.ts` file (entry point), just import it and then watch `appNeedsRefresh` and include: 

```typescript
const {
  appNeedsRefresh,
  offlineAppReady,
} = useServiceWorker()
watch(appNeedsRefresh, async() => {
  window.location.reload()
}, { immediate: true })
```

obviously, you will need to add some button on the screen, then show/enable it using `appNeedsRefresh` and then add an `@click` listener to this button to call `window.location.reload()`.

**Comment by scambier** [comment]
Thanks, that looks like a reasonable solution. I'll keep this ticket open until I implement this properly :)

**Comment by userquin** [comment]
@antfu As soon I have time I'll try to fix it using workbox instead register-service-worker, I'm very busy at work...

In the meantime this is the approach I'm using on my projects with vitesse template.

**Comment by userquin** [comment]
@antfu I have a problem about using `workbox-util`, let me explain:

- I'm trying use `Workbox` and `messageSW` to register service worker instead using navigation capabilities.
- I need to do this: [Offer a page reload for users](https://developers.google.com/web/tools/workbox/guides/advanced-recipes#offer_a_page_reload_for_users)

Then, the problem arises, I need to include `workbox-util` as a `dependency` not as a `dev dependency`: the problem is not here, is on the target project, for example, `vitesse template` where this plugin is installed as a `dev dependency`, and so, when building...

To put you in context:

My first attempt was to configure my pwa like this (`vite.config.ts`):

```typescript
    VitePWA({
      ...
      workbox: {
        cleanupOutdatedCaches: true,
        skipWaiting: true,
        clientsClaim: true,
      }
    }),
    ...
```

This aproach has a problem, the pages remains on the cache, also after a refresh and my pages stop working, because I removed the old registered service worker (`cleanupOutdatedCaches`): refresing page with F5 in online mode, request are served from the cache, so the oldest assets are missing from my server and the page stop working.

My second attemp was just removing `cleanupOutdatedCaches`, this will at least keep my pages working, but with oldest assets!!!.

The problem with activating `skipWaiting` is that the pages remains with the original assets links, and there is no way no bypass it.

The solution described in [Offer a page reload for users](https://developers.google.com/web/tools/workbox/guides/advanced-recipes#offer_a_page_reload_for_users) seems to work, but I haven't test it yet.

[In this link](https://medium.com/google-developer-experts/workbox-4-implementing-refresh-to-update-version-flow-using-the-workbox-window-module-41284967e79c) you can find the problem and how can be solved, in fact it points to the [Offer a page reload for users](https://developers.google.com/web/tools/workbox/guides/advanced-recipes#offer_a_page_reload_for_users), just read it.

We can also find a warning in workbox using `skipWaiting` [here](https://developers.google.com/web/fundamentals/instant-and-offline/service-worker/lifecycle#skip_the_waiting_phase)

**Comment by antfu** [comment]
Thanks a lot for the detailed info!

> Offer a page reload for users

This looks like a good solution to me! We can make an option for people to opt-in with this behavior.

> cleanupOutdatedCaches & skipWaiting

Do we need these tho?

**Comment by userquin** [comment]
No, just to remove both and do it manually with the `Offer a page reload for users` script: the script (module) included there is what we need to create here

**Comment by userquin** [comment]
this script will replace manual registration: what this plugin currently does will be replace with the script in `Offer a page reload for users`

**Comment by userquin** [comment]
What I had in mind was to include `useServiceWorker` in `@vueuse` using `workbox-util` and include the logic described in `Offer a page reload for users` with a callback: `createUIPrompt` would be an option to `useServiceWorker`. 

And then, here is the problem again...

**Comment by antfu** [comment]
I think maybe we don't need to be Vue specific (this plugin is framework-agnostic).

We could update this https://github.com/antfu/vite-plugin-pwa/blob/c2054640b1a81650a8172e1fe1dadb28178e4957/src/html.ts#L9-L16 to put the snippet from `Offer a page reload for users` and ask users to install `workbox-window` if they want to have this feature.

After that, we can provide a minimal example of how to set it up for people to explore. (and Vitesse as well for sure)

**Comment by userquin** [comment]
The problem with this approach is that is not tied to the app ui, that is, we need to add a callback, for example, with my first approach using `register-service-worker` using `appNeedsRefresh`.

I think it would be nice to add the `useServiceWorker` here with the logic, exposing this 2 guys (`appNeedsRefresh` and `offlineAppReady`, and expose some wrapper callback to be called from the ui once the user click on the refresh option.

**Comment by userquin** [comment]
As you can see the problem is mixing again buildtime and runtime, we need to interact with the ui to activated the service worker.

**Comment by userquin** [comment]
My suggestion is to include a new `custom` option and generate `useServiceWorker.ts`, but where to put it?. Then instruct/ask the user to include `workbox-build`/`workbox-window` as dependency and show how to use it (something similar in my  response to @scambier).

**Comment by antfu** [comment]
I am thinking about we serve the register script in a virtual module, so the usage would be like this:


```ts
// virtual module
import registerSW from 'vite-plugin-pwa-register'

const sw = registerSW({
  onNeedRefresh() {
    // show an UI for user to refresh
  }
})
```

**Comment by userquin** [comment]
wait to @scambier response, I think it must work (if not using `skipWaiting` , `clientsClaim` and `cleanupOutdatedCaches`)...

**Comment by scambier** [comment]
@userquin 

> You can use [register-service-worker](https://github.com/yyx990803/register-service-worker), and do a manual installation of service worker via `VitePWA` configuration on `vite.config.ts` file, something like this:
> ...

The code you proposed actually doesn't solve my issue it seems. In the end it just triggers a `location.reload()`, but a simple refresh isn't enough. Even a `ctrl+f5` with the cache disabled in the network tab doesn't work, I have to unregister the service worker to finally get the latest files.

**Comment by userquin** [comment]
@scambier yes, this is the problem I have, but with my SPA acting as a MPA (my server build route pages instead just returning/forwarding to index.html), and I havenn't tested on a real SPA, your answer confirms my suspicions...

can you try this one (just change the callback)?:

```typescript
updated: async(registration) => { 
  registration.update()
  appNeedsRefresh.value = true
},
```

@antfu try to make a branch where we can test it before merging into master, the problem is I haven't time...

**Comment by userquin** [comment]
upps, add `await` keyword: `await registration.update()`

**Comment by userquin** [comment]
@scambier the problem is that the service worker is installed but not activated, we need to provide a callback to the service worker to control the opened pages/tabs: the lifecycle of the service worker is a little complicated. 

What I'm talking with @antfu is to try to activate the service worker once is updated, but we need to change the code. 

For example, vuejs 3 (using workbox 4, here we are using 6.1.1), has some code, I'll try to investigate, to activate and claim clients to take control of opened windows/tabs controlled by the service worker (I need to see where attaching the listener to show the dialog for `New content available`):

`from https://v3.vuejs.org/ => at the top of the service-worker.js file`
```javascript
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'SKIP_WAITING') {
    self.skipWaiting();
  }
});
```

`from https://v3.vuejs.org/ => at the bottom of the service-worker.js file`
```javascript
addEventListener('message', event => {
  const replyPort = event.ports[0]
  const message = event.data
  if (replyPort && message && message.type === 'skip-waiting') {
    event.waitUntil(
      self.skipWaiting().then(
        () => replyPort.postMessage({ error: null }),
        error => replyPort.postMessage({ error })
      )
    )
  }
})
```

**Comment by userquin** [comment]
@antfu vuejs next docs is using `@vuepress/pwa` and [here](https://github.com/vuejs/vuepress/blob/master/packages/%40vuepress/plugin-pwa/index.js) is the code generating the `service-worker.js` file (similiar to this one but touching also the app entry). 

It uses a global registered component to show the dialog on new content available.

**Comment by userquin** [comment]
Following the code, it seems using `Offer a page reload for users` can simplify the logic on `@vuepress/pwa`: see lib directory.

**Comment by scambier** [comment]
> @scambier yes, this is the problem I have, but with my SPA acting as a MPA (my server build route pages instead just returning/forwarding to index.html), and I havenn't tested on a real SPA, your answer confirms my suspicions...
> 
> can you try this one (just change the callback)?:
> 
> ```ts
> updated: async(registration) => { 
>   registration.update()
>   appNeedsRefresh.value = true
> },
> ```
> 
> @antfu try to make a branch where we can test it before merging into master, the problem is I haven't time...

This seems to do the job.

Here's what I have at the moment:

useServiceWorker.ts
```ts
register(
    '/sw.js',
    {
      registrationOptions: { scope: '/' },
      updated: async (registration) => {
        await registration.update()
        registration.unregister()
        appNeedsRefresh.value = true
      },
      cached: () => {
        offlineAppReady.value = true
      }
    }
  )
```

main.ts
```ts
  const { appNeedsRefresh, offlineAppReady } = useServiceWorker()
  watch(appNeedsRefresh, async (val) => {
    if (val) {
      console.log('app updated and sw unregistered, will refresh')
      // TODO: prompt user
      location.reload()
    }
  }, { immediate: true })
```

There's just a point I'm missing. In your first response, you wrote
> do a manual installation of service worker via VitePWA configuration

But I'm not quite sure which setting to change.

**Comment by userquin** [comment]
`vite.config.ts`
```typescript
VitePWA({
  injectRegister: null,
})
```

**Comment by userquin** [comment]
to clear the cache use this:

`vite.config.ts`
```
VitePWA({
  ...
  workbox: {
    cleanupOutdatedCaches: true,
  }
}),

**Comment by userquin** [comment]
if you don't have `injectRegister: null` you're registering the service worker twice, former in the html and second one in `maint.ts`.

Setting `injectRegister: null` will just remove the registration of the sw from the index.html file (see current index.html and you will see `navigator.serviceWorker.register` or a script tag pointing to `registerSW.js`).

And please, confirm that update call just works or adding unregister.

**Comment by userquin** [comment]
Testing on my MPA it seems we need to unregister first, then update and then refresh page, but still service worker not controlling page/tabs.

About the cache @scambier see screenshot below.

I'm checking if `clientsClain: true,` will do the work...

@antfu With this approach there is no need to modify code, just wait until my test can confirm that works...

![imagen](https://user-images.githubusercontent.com/6311119/110133272-6a96cc00-7dcc-11eb-86ec-6dc58138b502.png)

**Comment by scambier** [comment]
Ok, the app correctly cleans the cache after a new deployment, but _only_ if I refresh or open a new tab. And in that case, it only reloads the first tab. The `updated: async (registration) => {}` never triggers by itself after a new deployment.

To recap, here's what I currently have.

useServiceWorker.ts:
```ts
import type { Ref } from 'vue'

import { register } from 'register-service-worker'
import { ref } from 'vue'

// https://medium.com/google-developer-experts/workbox-4-implementing-refresh-to-update-version-flow-using-the-workbox-window-module-41284967e79c
// https://github.com/yyx990803/register-service-worker
// https://developers.google.com/web/fundamentals/primers/service-workers/high-performance-loading
export const useServiceWorker = (): {
  appNeedsRefresh: Ref<boolean>
  offlineAppReady: Ref<boolean>
} => {
  const offlineAppReady = ref(false)
  const appNeedsRefresh = ref(false)

  register(
    '/sw.js',
    {
      registrationOptions: { scope: '/' },
      updated: async (registration) => {
        await registration.update()
        registration.unregister()
        appNeedsRefresh.value = true
      },
      cached: () => {
        offlineAppReady.value = true
      }
    }
  )

  return {
    appNeedsRefresh,
    offlineAppReady
  }
}
```

main.ts (after `createApp(App)`):

```ts
  const { appNeedsRefresh } = useServiceWorker()
  watch(appNeedsRefresh, async (val) => {
    console.log('Does app need refresh? ' + val)
    if (val) {
      // if (confirm('App updated. Do you want to refresh?')) {
      location.reload()
      // }
    }
  }, { immediate: true })
```

PWA config in vite.config.ts:

```ts
VitePWA({
      injectRegister: null,
      manifest: {
        /* */
      },
      workbox: {
        cleanupOutdatedCaches: true
      }

    })
```

Thanks a lot for your explanations. I'll take time to read some more doc about service workers next week :)

**Comment by userquin** [comment]
I'm trying to simulate Workbox approach, just copy/paste `workbox-window` into my MPA: if working I'll provide the virtual module to be used...

**Comment by userquin** [comment]
@antfu confirmed that works at least on my MPA with the `/` route, below the virtual module.

Just update the imports for `workbox-window`, remove the import for `useRouter` and remove the `router.isReady()` callback and remove all unnecesary comments.

The usage is calling `updateServiceWorker` instead reload the window once the user click on the UI once `appNeedsRefresh` is activated.

```typescript
import type { Ref } from 'vue'

import { Workbox } from '/~/logics/service-worker/Workbox'
import { messageSW } from '/~/logics/service-worker/messageSW'

import { ref } from 'vue'
import { useRouter } from 'vue-router'

export const useServiceWorker = (immediate = false): {
  offlineAppReady: Ref<boolean>
  appNeedsRefresh: Ref<boolean>
  updateServiceWorker: () => Promise<void>
} => {
  const offlineAppReady = ref(false)
  const appNeedsRefresh = ref(false)
  const router = useRouter()

  let registration: ServiceWorkerRegistration
  let wb: Workbox

  const updateServiceWorker = async() => {
    // Assuming the user accepted the update, set up a listener
    // that will reload the page as soon as the previously waiting
    // service worker has taken control.
    wb.addEventListener('controlling', (event) => {
      if (event.isUpdate)
        window.location.reload()
    })

    if (registration && registration.waiting) {
      // Send a message to the waiting service worker,
      // instructing it to activate.
      // Note: for this to work, you have to add a message
      // listener in your service worker. See below.
      await messageSW(registration.waiting, { type: 'SKIP_WAITING' })
    }
  }

  router.isReady().then(() => {
    if ('serviceWorker' in navigator) {
      wb = new Workbox('/sw.js', { scope: '/' })

      const showSkipWaitingPrompt = () => {
        // `event.wasWaitingBeforeRegister` will be false if this is
        // the first time the updated service worker is waiting.
        // When `event.wasWaitingBeforeRegister` is true, a previously
        // updated service worker is still waiting.
        // You may want to customize the UI prompt accordingly.

        // Assumes your app has some sort of prompt UI element
        // that a user can either accept or reject.
        appNeedsRefresh.value = true
      }

      wb.addEventListener('controlling', (event) => {
        if (!event.isUpdate)
          offlineAppReady.value = true
      })
      // Add an event listener to detect when the registered
      // service worker has installed but is waiting to activate.
      wb.addEventListener('waiting', showSkipWaitingPrompt)
      // @ts-ignore
      wb.addEventListener('externalwaiting', showSkipWaitingPrompt)

      wb.register({ immediate }).then(r => registration = r!)
    }
  })

  return {
    offlineAppReady,
    appNeedsRefresh,
    updateServiceWorker,
  }
}
```

**Comment by userquin** [comment]
after a revision:

```typescript
import type { Ref } from 'vue'

import { Workbox } from '/~/logics/service-worker/Workbox'
import { messageSW } from '/~/logics/service-worker/messageSW'

import { ref } from 'vue'
import { useRouter } from 'vue-router'

export const useServiceWorker = (immediate = false): {
  offlineAppReady: Ref<boolean>
  appNeedsRefresh: Ref<boolean>
  updateServiceWorker: () => Promise<void>
} => {
  const offlineAppReady = ref(false)
  const appNeedsRefresh = ref(false)
  const router = useRouter()

  let registration: ServiceWorkerRegistration

  const updateServiceWorker = async() => {
    if (registration && registration.waiting) {
      // Send a message to the waiting service worker,
      // instructing it to activate.
      // Note: for this to work, you have to add a message
      // listener in your service worker. See below.
      await messageSW(registration.waiting, { type: 'SKIP_WAITING' })
    }
  }

  router.isReady().then(() => {
    if ('serviceWorker' in navigator) {
      const wb = new Workbox('/sw.js', { scope: '/' })

      const showSkipWaitingPrompt = () => {
        // `event.wasWaitingBeforeRegister` will be false if this is
        // the first time the updated service worker is waiting.
        // When `event.wasWaitingBeforeRegister` is true, a previously
        // updated service worker is still waiting.
        // You may want to customize the UI prompt accordingly.

        // Assumes your app has some sort of prompt UI element
        // that a user can either accept or reject.
        appNeedsRefresh.value = true
      }

      wb.addEventListener('controlling', (event) => {
        if (event.isUpdate)
          window.location.reload()
        else
          // Assuming the user accepted the update, set up a listener
          // that will reload the page as soon as the previously waiting
          // service worker has taken control.
          offlineAppReady.value = true
      })
      // Add an event listener to detect when the registered
      // service worker has installed but is waiting to activate.
      wb.addEventListener('waiting', showSkipWaitingPrompt)
      // @ts-ignore
      wb.addEventListener('externalwaiting', showSkipWaitingPrompt)

      wb.register({ immediate }).then(r => registration = r!)
    }
  })

  return {
    offlineAppReady,
    appNeedsRefresh,
    updateServiceWorker,
  }
}
```

**Comment by userquin** [comment]
also working on my MPA, just configuring `templatesURL` on `workbox` including all dynamic routes generated on server side.

The configuration for `VitePWA` will be as @scambier reported:

```typescript
VitePWA({
      injectRegister: null,
      manifest: {
        /* */
      },
      workbox: {
        cleanupOutdatedCaches: true
      }
})
```

**Comment by userquin** [comment]
@antfu I attach here the `copy/paste` files from workbox 6.0 that are the same for 6.1.1 if you want to play with it before starting...

[service-worker.zip](https://github.com/antfu/vite-plugin-pwa/files/6092316/service-worker.zip)

**Comment by userquin** [comment]
@scambier can you just download the zip from previous comment, unzip it on your src/logics directory (create it if does not exist) and change the logic to import this new one `useServiceWorker` calling `updateServiceWorker` method from the button?

**Comment by userquin** [comment]
I upload here new service-worker, second approach doesn't work, we need to register on right registration!!!

[service-worker.zip](https://github.com/antfu/vite-plugin-pwa/files/6093016/service-worker.zip)

**Comment by userquin** [comment]
@antfu ping

**Comment by antfu** [comment]
Thanks for looking into it. But I am afraid if we need that complicated approach, and wish we are not coupled with Vue. Maybe I will take some time to give a shot later.

**Comment by userquin** [comment]
This is an example for my tests. We can provide some virtual module
abstracting callbacks and workbox things.

Try to see logic on vuepress....

**Comment by userquin** [comment]
The problem is that if you dont provide this, the pwa just stop working, see the title of this issue...

**Comment by userquin** [comment]
One last thing to change about the `ready to work offline message`, we need to listen to `activated` event instead `controlling` one:

```typescript
      wb.addEventListener('activated', (event) => {
        // this will only controls the offline request.
        // `event.isUpdate` will be true if another version of the service
        // worker was controlling the page when this version was registered.
        if (!event.isUpdate)
          offlineAppReady.value = true
      })
```

A proporsal will be somethig like this:

```typescript
export type RegisterSWOptions = {
  scriptURL?: string
  scope?: string
  immediate?: boolean
  onNeedRefresh: () => void
  onOfflineReady?: () => void
}

export type UpdateServiceWorkerFn = () => Promise<void>

export declare function registerSW(registerSWOptions: RegisterSWOptions): UpdateServiceWorkerFn

```

the implementation will be something like this:

```typescript
import { Workbox, messageSW } from 'workbox-window'

export const registerSW = (registerSWOptions: RegisterSWOptions): UpdateServiceWorkerFn => {

  let registration: ServiceWorkerRegistration
  let wb: Workbox

  const updateServiceWorker = async() => {
    // Assuming the user accepted the update, set up a listener
    // that will reload the page as soon as the previously waiting
    // service worker has taken control.
    wb && wb.addEventListener('controlling', (event) => {
      if (event.isUpdate)
        window.location.reload()
    })
    if (registration && registration.waiting) {
      // Send a message to the waiting service worker,
      // instructing it to activate.
      // Note: for this to work, you have to add a message
      // listener in your service worker. See below.
      await messageSW(registration.waiting, { type: 'SKIP_WAITING' })
    }
  }

  if ('serviceWorker' in navigator) {
    wb = new Workbox(registerSWOptions.scriptURL || '/sw.js', { scope: registerSWOptions.scope || '/' })

    const showSkipWaitingPrompt = () => {
      // `event.wasWaitingBeforeRegister` will be false if this is
      // the first time the updated service worker is waiting.
      // When `event.wasWaitingBeforeRegister` is true, a previously
      // updated service worker is still waiting.
      // You may want to customize the UI prompt accordingly.

      // Assumes your app has some sort of prompt UI element
      // that a user can either accept or reject.
      registerSWOptions.onNeedRefresh()
    }

    wb.addEventListener('activated', (event) => {
      // this will only controls the offline request.
      // `event.isUpdate` will be true if another version of the service
      // worker was controlling the page when this version was registered.
      if (!event.isUpdate && registerSWOptions.onOfflineReady)
        registerSWOptions.onOfflineReady()
    })
    // Add an event listener to detect when the registered
    // service worker has installed but is waiting to activate.
    wb.addEventListener('waiting', showSkipWaitingPrompt)
    // @ts-ignore
    wb.addEventListener('externalwaiting', showSkipWaitingPrompt)
    // register the service worker
    wb.register({ immediate: registerSWOptions.immediate || false }).then(r => registration = r!)
  }

  return updateServiceWorker
}
```

And so, for `vue` (usage) we can have something like this:
```typescript
import registerSW from 'vite-plugin-pwa-register'

export const useServiceWorker = (immediate = false, scriptURL = '/sw.js', scope = '/'): {
  offlineAppReady: Ref<boolean>
  appNeedsRefresh: Ref<boolean>
  updateServiceWorker: () => Promise<void>
} => {
  const offlineAppReady = ref(false)
  const appNeedsRefresh = ref(false)

  const updateServiceWorker = registerSW({
    scriptURL,
    scope,
    immediate,
    onNeedRefresh: () => {
      appNeedsRefresh.value = true
    },
    onOfflineReady: () => {
      offlineAppReady.value = true
    },
  })
  
  return {
    offlineAppReady,
    appNeedsRefresh,
    updateServiceWorker,
  }
}
```

**Comment by userquin** [comment]
@antfu we need that complicated approach just because of lifecycle of the service worker , there is no another way to do it...

If you look at  `@vuepress/pwa`, that is scary: modify the script generated by workbox-build, also modifying the ui and register listeners ...

**Comment by antfu** [comment]
Wow @userquin, that's exactly what I was thinking about. Great work!

In the addition of 

```ts
import registerSW from 'vite-plugin-pwa-register'
```

When we could have something like 

```ts
import { useServiceWorker } from 'vite-plugin-pwa-register/vue'
import { useServiceWorker } from 'vite-plugin-pwa-register/react'
```

as the extensions.

**Comment by userquin** [comment]
It only remains to make it nice and unify the types a little

I started to make the virtual module in the plugin, but I mess with the workspaces, and how to structure everything within the project, although I think it should be outside

**Comment by antfu** [comment]
@userquin Just sent you the invitation to this project. Appreciated all you have done on this topic!

Please feel free to create a new branch and draft a PR, I am happy to review and help with the discussion and implementation! Thanks

**Comment by userquin** [comment]
@antfu a few questions:

1) How to start it? Make a new folder for virtual module. Where to put dependencies?
2) How about extensions? I don't have any idea how to begin
3) About react I haven't used it, only vue and some little things with svelte.

Maybe you can add a new branch where I can start (clone it), at least with the structure created...

**Comment by antfu** [comment]
Re: 1, serving virtual module is actually quite simple, you can see the docs here:
https://vitejs.dev/guide/api-plugin.html#importing-a-virtual-file

Re: 2, they are just like other virtual modules

Re: 3, no pressure on that, we can leave them for the community to contribute, or if you want, you can do the vue part for sure.

**Comment by userquin** [comment]
@antfu I have a few problems with virtual modules, I cannot use typescript, just plain javascript.

I'm testing example module and I'll make an initial push to my fork.

I have gotten it to at least compile:

`example App.vue`
```html
<script lang="ts">
import { defineComponent, ref } from 'vue'
import { registerSW } from 'vite-plugin-pwa-register'
export default defineComponent({
  setup() {
    const offlineAppReady = ref(false)
    const appNeedsRefresh = ref(false)
    const { updateServiceWorker } = registerSW({
      immediate: false,
      onNeedRefresh: () => {
        appNeedsRefresh.value = true
      },
      onOfflineReady: () => {
        offlineAppReady.value = true
      },
    })
    return {
      offlineAppReady,
      appNeedsRefresh,
      updateServiceWorker,
    }
  },
})
</script>
<template>
  <div v-if="offlineAppReady">Ready to work offline</div>
  <div v-if="appNeedsRefresh">
    New content available
    <button @click="updateServiceWorker">Refresh</button>
  </div>
  <div>Hello World</div>
</template>
```
`virtual module`
```javascript
import { Workbox, messageSW } from 'workbox-window'

export function registerSW(
  immediate,
  onNeedRefresh,
  onOfflineReady,
) {

  let registration
  let wb

  function updateServiceWorker() {
    // Assuming the user accepted the update, set up a listener
    // that will reload the page as soon as the previously waiting
    // service worker has taken control.
    wb && wb.addEventListener('controlling', function(event) {
      if (event.isUpdate)
        window.location.reload()
    })
    if (registration && registration.waiting) {
      // Send a message to the waiting service worker,
      // instructing it to activate.
      // Note: for this to work, you have to add a message
      // listener in your service worker. See below.
      messageSW(registration.waiting, { type: 'SKIP_WAITING' })
    }
  }

  if ('serviceWorker' in navigator) {
    wb = new Workbox('/sw.js', { scope: '/' })

    wb.addEventListener('activated', function(event) {
      // this will only controls the offline request.
      // event.isUpdate will be true if another version of the service
      // worker was controlling the page when this version was registered.
      if (!event.isUpdate && typeof onOfflineReady === 'function')
        onOfflineReady()
    })
    // Add an event listener to detect when the registered
    // service worker has installed but is waiting to activate.
    wb.addEventListener('waiting', onNeedRefresh)
    // @ts-ignore
    wb.addEventListener('externalwaiting', onNeedRefresh)
    // register the service worker
    wb.register({ immediate }).then(r => registration = r)
  }

  return updateServiceWorker
}
```

**Comment by antfu** [comment]
Great, looking forward to it! Yeah, you should use plain js + .d.ts to make it work. We can still write them in ts and transpile them into js in the dist to serve. If you don't get what I mean, you can leave the ts version in the source code and I can take care of the build setup

**Comment by userquin** [comment]
ok, later I'll try to have both variants, one commented, just coment/remove what you want.

For you info:

1) added modules.ts, just to create the virtual module(s): currently only plain supported (we need to add vue and react: I'll try to add former)
2) modify index.ts to include `virtualFileId` and added  `resolveId` and `load` hooks to revolve virtual modules
3) `App.vue`: just for testing
4) added `register` option on `injectRegister` option
5) added `workbox-window` as dependency

can you tell me your timezone? I'm from Spain, GMT + 1

**Comment by antfu** [comment]
UTC+8, and it's 2 AM for me now :P

I am going to have some sleep now, will get back to you later.

**Comment by userquin** [comment]
@antfu tested on example project and first push to my branch: https://github.com/userquin/vite-plugin-pwa

Thinks you must see:

1) added `src/modules.ts`: virtual modules with javascript (typescript on a second round). Also a first version to generate `vite-plugin-pwa-register/vue` virtual module.
2) modified `src/types.ts` to include `register` as another option for `injectRegister`.
3) modified `src/index.ts` to include `virtualFileId` and added `resolveId` and `load` hooks to revolve virtual modules.
4) added `workbox-window` as dependency to `package.json`. Only generated when `injectRegister: 'register``. We need to review current options.
5) modified `example/vite.config.ts` to remove outdated caches and changed `injectRegister` to `register`: just added `workbox:  { cleanupOutdatedCaches: true }`. Also modified `base`entry to allow bypass hardcoded `base: 'https://github.com/' and added `sourcemap`to allow see source on local tests: `build: { sourcemap: process.env.SOURCE_MAP === 'true', }` (see next entry, crossenv passed on script).
6) modified `example/package.json` to include `node-static` dependency to do local tests and added `run-build` script entry to `build + serve via node-static` example: I have added some `crossenv` variables to allow test on my local ip, just read it.
7) modified `package.json`: added `example:run-build` script to allow `build + run example`: `pnpm example:run-build` just build the example and run it on node via `st-pwa-with-node-static.js` (see 9).
8) modified `example/src/App.vue`: just for testing and with `vite-plugin-pwa-register`. A real example => how to configure it.
9) added `example/test-pwa-with-node-static.js` to test on `example` build with `node-static`.
10) `.gitignore` to exclude certificates and `.env.local.json`: see 2) bellow.

Anyway, go to my fork and just compare it with this to see changes I have made.

If you want to test it, you will need:

1) an ssl certificate, for example, use `tls-keygen`, then
2) add `example/.env.local.json` file: see `.env.local.json` and `.env.local.json.content` images bellow . You will need to configure `node-static` ssl stuff and optionally you can configure `hostname`, `https port` and `http port` (`localhost`, `443` and `80` respectivelly). Just see `example/test-pwa-with-node-static.js` script.
3) for first test (`app ready to work offline`, see `app-ready-offline image` bellow) you only need to execute `pnpm example:run-build`, open a browser with dev tools opened.
4) to test `New content available`, stop `node` script, then go to `example/src/Vue.app` and change `return 'App ready to work offline'` with `return 'Application ready to work offline'`, then run again `pnpm example:run-build` and go to page on browser opened on previous step, click on navigation bar and just press enter (), wait a few seconds and then press on refresh button (see `app-ready-offline image` bellow).

`.env.local.json image`
![.env.local.json](https://user-images.githubusercontent.com/6311119/110255693-4c2ffc80-7f95-11eb-9cc5-b10dcae93fe0.png)

`.env.local.json.content image`
![imagen](https://user-images.githubusercontent.com/6311119/110255770-af219380-7f95-11eb-86ff-7ec2541c83d2.png)

Here 2 animated gifs:

`app-ready-offline image`
![app-ready-offline](https://user-images.githubusercontent.com/6311119/110255390-ae87fd80-7f93-11eb-894f-d0ee7e1cff2d.gif)

`first test assets image`
![imagen](https://user-images.githubusercontent.com/6311119/110256398-02e1ac00-7f99-11eb-991b-3c472a2de690.png)

`first test assets directory image`
![imagen](https://user-images.githubusercontent.com/6311119/110256418-1ab93000-7f99-11eb-977d-0faa1b9e0c46.png)

`new content available image`
Sorry, animated GIF is too big => 10MB max. I just send you.

`new content available assets image`
![imagen](https://user-images.githubusercontent.com/6311119/110256651-412b9b00-7f9a-11eb-8017-835a5c8a1e20.png)

`new content available assets dir image`
![imagen](https://user-images.githubusercontent.com/6311119/110256481-6966ca00-7f99-11eb-8f00-0061f8693d79.png)

**Comment by userquin** [comment]
@antfu the animated gif also cannot be sent it is too big.

**Comment by userquin** [comment]
![new-content-available-2](https://user-images.githubusercontent.com/6311119/110256817-3de4df00-7f9b-11eb-9552-37c171e3c73e.gif)

**Comment by userquin** [comment]
@scambier @antfu here an animated gif to see the problem described in this issue solved with the new virtual module.

![ezgif com-video-to-gif(1)](https://user-images.githubusercontent.com/6311119/110258292-307f2300-7fa2-11eb-8268-389adea4ae60.gif)

**Comment by userquin** [comment]
@antfu added ssl certificate for `localhost`

you only need to add this content on `example/.env.local.json` file:

```json
{
  "hostname": "localhost",
  "ca": "localhost-cert.pem",
  "cer": "localhost-cert.pem",
  "key": "localhost-key.pem"
}
```

**Comment by userquin** [comment]
I think we need to add a test with `vue-router` (at least 1 more route, for example `about`) to confirm that works, I just test with `/` and we need to confirm that another routes get update also with its assets.

My test with my MPA is working with `vue-router`, but still with `vite and vitesse v1`.

**Comment by scambier** [comment]
Hello, it seems you had a productive weekend ^^

It's not really clear from your gifs though, does the "new content available" trigger by itself, without first requiring a manual reload of the app?

**Comment by antfu** [comment]
Wow! @userquin thanks for so much work you have done! Please let me take some time to go through it. Instead of doing it in the fork, please do that on the main repo so we can better collaborate. I have created https://github.com/antfu/vite-plugin-pwa/tree/feat/sw-enhancement branch and forward your commits to it. And I will do the review and changes on it as well. Thanks a lot!

**Comment by userquin** [comment]
@scambier 

- the former animated gif is just entering url on browser and press enter, then the toast appears without nothing more first time the service worker is installed
- second animated gif just shows previous page, clicking on navigation bar then pressing enter (before that I just make a change on App.vue to create a new version for at least index.js asset): after a few seconds (miliseconds, depends on the size of the app), the toast appears and on the left panel (devtools), the new service worker appears with state to skip awaiting below the old one (the green flag is the old service worker, the yellow one is for this new version with skip waiting state). Once refresh button clicked (you can see the tab reload spinner once clicked), then old service worker is removed, caches updated and new service worker activated (sorry for the quality of the gifs, I haven't found a way to create them bellow 10MB).

@antfu

I just awaiting to you, I'll try to push a PR to this branch new branch: I'm very happy, now I can work directly on the repo, saving time each time is updated.

**Comment by userquin** [comment]
![imagen](https://user-images.githubusercontent.com/6311119/110303102-491b2780-7ffa-11eb-810e-0bce51103ccd.png)

**Comment by userquin** [comment]
![imagen](https://user-images.githubusercontent.com/6311119/110303208-694ae680-7ffa-11eb-8dc6-8cc85e59a0a9.png)

**Comment by userquin** [comment]
@scambier screenshots for previous descripcion take on the captured video: on last image you can see the spinner on the tab: sorry my edge browser is on spanish (at least the states in english).

**Comment by userquin** [comment]
@antfu Have you make the PR to the new branch?

**Comment by userquin** [comment]
@antfu ok, I misunderstood your comment, you have already updated the new branch with my commits.

**Comment by userquin** [comment]
@antfu A question about virtual modules: I work with IntelliJ (that needs some plugin rework to support setup script and things like virtual modules). 

Do VSCode resolve this virtual modules?

![imagen](https://user-images.githubusercontent.com/6311119/110305031-8bddff00-7ffc-11eb-9901-223cb539a45a.png)

**Comment by antfu** [comment]
@userquin man, I'd say it working really well. Thanks again. I have made a few changes to the branch, please take a look

- You can now use TS to write code for the generated code, the build process will transpile it to js
- .d.ts solved.
- A few changes to the example, now every time you build, the dist will be default as it contains the current time info for you to check if the reload is successful.

I am going to take a break for now, but generally, it's pretty good, awesome work 👍

Will merge soon after a few more overview

**Comment by scambier** [comment]
> * second animated gif just shows previous page, clicking on navigation bar then pressing enter (before that I just make a change on App.vue to create a new version for at least index.js asset

@userquin Alright, so you still need to do a manual reload so that it can detect that app has been updated, right? I understand that you generalized everything to get that inside the plugin, but I thought (erroneously?) that the worker would be able to detect itself that `sw.js` had been updated without needing a reload.

**Comment by userquin** [comment]
@scambier YES, the user has to take the control about it, you cannot refresh automatically: all articles I have read from google stuff explains that you cannot do it automatically (a user filling a form can lost the data).

About to self detection can be done with push notifications.

Usually, the detection is when the user re-enters to the a page (that triggers a "fresh" request), pwa installed or not.

**Comment by userquin** [comment]
@antfu your changes look awesome, I'm using typescript for 3 or 4 months so I'm very new...

**Comment by scambier** [comment]
> YES, the user has to take the control about it, you cannot refresh automatically: all articles I have read from google stuff explains that you cannot do it automatically (a user filling a form can lost the data).

There's a misunderstanding here. I'm asking if the prompt "_new content, please refresh_" will show itself without user intervention, if the worker can detect that `sw.js` has been updated so the prompt needs to be shown.

Because if you need to refresh the app to trigger showing the prompt, what's the point of it? Sorry if the answer is obvious, but I'm having a hard time following this whole thread.

**Comment by userquin** [comment]
No: always the user intervention is required. Instead waiting the user open again any webpage of your site (or press F5), from your server you can send a push notification to update the service worker, just like if the user open the browser again o press F5.

But this is another history, you will need to suscribe the client on server side to this notifications...

A service worker needs the browser to be opened, and the user in any webpage of your site.

There is an exception on mobile devices and push notifications, where the operating system (push) notifications is used, but again, user interaction is required and when PWA is installed (I haven't tested).

**Comment by scambier** [comment]
If I'm getting your explanations right, I need to reload the app so that it can show me a prompt asking me to reload. So what's the point? I'm already reloading the app, I might as well clear the cache while I'm at it, instead of asking for another refresh.

**Comment by userquin** [comment]
You describe a test scenario. Users just visit your page, then shutdown the laptop, or just hide the browser on mobile. If you want to avoid your users to reload pages to see if there is a new version, you just configure push notifications on your server side and then suscribe to it your website.  Then the user don't need to refresh to see if the is a new version, the server will notify them when a new version arrives.

Anyway, this issue is about how to configure the service worker to work properly on an update, via F5, via opening again the browser via push notifications to solve the cache problem. It is not about user interaction, both are related, but not the problem to solve.

**Comment by scambier** [comment]
Ok, then [the issue is solved since Friday](https://github.com/antfu/vite-plugin-pwa/issues/33#issuecomment-791448708), or is there something obvious I'm missing?

**Comment by userquin** [comment]
@scambier 

No, the issue is being resolved, @antfu has create PR #34 . Once merged on main it will be resolved also this issue ()do not close it, github will close automatically).

Anyway, read service worker documentation, maybe your app cannot be used with a service worker, you need to review if matches your requisites.

For example, my MPA was originally to be 2 native apps to work on android and ios. The resources spent to do this compared to make it a PWA are significant. My decision was to make a PWA and so I need a service worker, it also needs to be able to work offline like natives ones. I have to include in my PWA server push notifications to avoid what you are describing...

If your server doesn't have push notifications, either, remove PWA or just make and alternate api entry point to test if there is a new version. For example, make some configuration, a timestamp an internal version number and store it on local storage. You can periodically check it to some api serving a plain number and simulate a refresh.

**Comment by userquin** [comment]
@scambier https://github.com/antfu/vite-plugin-pwa/pull/34/commits/15f51070f4e235e76a51d8ba8484b783e296a4e9

**Comment by MartinMalinda** [comment]
I tried to read through the discussion here and I'm afraid I couldn't really digest it all :d So a quick question:

Couldn't I just have a `networkFirst` strategy on the main index.html file? Or that is not allowed from the PWA point of view? The html file on `/` simply has to be delivered from cache no matter what?

**Comment by userquin** [comment]
@posva welcome to the hell: the main problem is bypass the broser cache and then its own cache: the new version will use StaleWhileRevalidate strategy, that means, go to cache and then in background go to server and then update if necessary.

If we go NetworkFirst strategy then we don't need the service worker, only for offline requirement. In some situations this will work but as a general solution to speed up with local cache then is not suitable.

As a mention above, you will need to check if the app requisites matches the usage of a service worker.

In fact, I have some apps that runs below 200ms using server push with esmodules for any page, so if I configure it with NetworkFirst strategy, the service worker will only be necessary if offline is required.

About the user click on the button, it is a general adoption, just see vuejs docs, vuetify docs, nuxt docs (vuepress)...

I'm talking with @antfu and have included a new `auto` option to bypass any user interaction.

**Comment by userquin** [comment]
@posva Creo que hablas español, te lo explico a ver si resuelve tus dudas.

El problema con los service workers es que tiene un ciclo de vida bastante entrerrevesado y por eso hay cosas que no encajan o no cuadran bien.

El principal problema está en el lapso de tiempo entre que un service worker está esperando a ser activado (skip waiting) y el momento en que se activa (activated). Esto que parece tan simple es un problema ya que provoca que no puedas hacer el bypass de la caché del mismo selectivamente.

Mientras un service worker no está activado, entonces es como si no existiese desde el punto de vista del aplicativo. El que incluyas una estrategia u otra no importa, porque lo que no tienes en cuenta es que quien realmente te está retornando lo que solicitas al backend es el que está activo.

Cuando has planteado la duda, has obviado este matiz, ya que aunque vayas al backend a por los datos y te retorne los últimos, te va a funcionar en ese instante, pero si haces un F5, te lo va a retornar el service worker activo, recuerda que todavía el nuevo no está activado, con lo que solicitará los assets con la versión anterior y lo que pasará es que se te quedará la pantalla en blanco y si miras la pestaña de red en el devtools, verás 404 de todos los recursos antiguos.

Como ves no es que falle en un punto o en otro, es un compendio de cosas, esta incidencia refleja lo que te he descrito y para llegar a ello he estado unos cuantos meses dándoles vueltas a cosas similares a tu planteamiento y que realmente no solucionaban nada.

Si estás un poco sorprendido, creo que hago mención a lo que hace `@vuepress/pwa` para poder sacar el `Nuevo contenido disponible`: básicamente espera en el build a que `workbox-build` termine y luego lo vuelve a modificar.

Si miras este comentario https://github.com/antfu/vite-plugin-pwa/pull/34#issuecomment-792798440, se lo explico a Anthony, y aún así no consiguíamos hacer el bypass de la caché del navegador.

**Comment by posva** [comment]
@userquin creo que te equivocaste de ping 😅 
Buena suerte!

**Comment by userquin** [comment]
Leches pues ni me he dado cuenta, perdona Eduardo por las molestias, juraría que escribí el nombre bien al responder, habrá sido el autocomplete de las narices... como el predictivo del telefono

**Comment by MartinMalinda** [comment]
Nuxt seems to also go in this direction: https://github.com/nuxt-community/pwa-module/issues/406

**Comment by userquin** [comment]
I forgot to mention that the current version is using `workbox.runtimeCaching`, you can see it on `src/cache.ts`, in #34 this file has been removed in favor of `precaching` via `precacheAndRoute`. The strategies used in the script are using `NetworkFirst` for `html`, and `StaleWhileRevalidate` for `sw.js` and `workbox-**.js` files (for `.js` files in general), but the problem still persists.

**Comment by userquin** [comment]
@MartinMalinda following your links, I think we have the same behavior described in https://github.com/GoogleChrome/workbox/issues/1767#issuecomment-729495174

What the new PR #34  generates is this:
```javascript
  self.addEventListener('message', event => {
    if (event.data && event.data.type === 'SKIP_WAITING') {
      self.skipWaiting();
    }
  });
  /**
   * The precacheAndRoute() method efficiently caches and responds to
   * requests for URLs in the manifest.
   * See https://goo.gl/S9QRab
   */

  workbox.precacheAndRoute([{
    "url": "assets/index.40de6d3d.css",
    "revision": "0b973314e89c1b1876dfd49f6241cdde"
  }, {
    "url": "assets/index.946e44f0.js",
    "revision": "6c4f2bbf84081cacf649a67aa3f28b72"
  }, {
    "url": "assets/vendor.35599eb4.js",
    "revision": "57c102d464ad82af5c1c300d2de100a0"
  }, {
    "url": "index.html",
    "revision": "f4d735ef53aae1144d2a255bc7d8bf9b"
  }], {});
  workbox.cleanupOutdatedCaches();
  workbox.registerRoute(new workbox.NavigationRoute(workbox.createHandlerBoundToURL("index.html")));
```

I will test if enabling `skipWaiting` and `clientClaims` on `workbox` entry of `VitePWA` has the effect you want: enabling this 2 flags then, `workbox-build` will replace:
```javascript
 self.addEventListener('message', event => {
    if (event.data && event.data.type === 'SKIP_WAITING') {
      self.skipWaiting();
    }
  });
```
with:
```javascript
  self.skipWaiting();
  workbox.clientsClaim();
```
I need to check if both are equivalent.
The downside with this approach will be we need to also add `NetworkFirst` to all entry pages when using `SSG` (for example with vitesse template).

**Comment by userquin** [comment]
We have this:

![imagen](https://user-images.githubusercontent.com/6311119/110618656-3ee26000-8197-11eb-8f87-ff987fa57730.png)

Pressing F5 or pressing enter on url box (see the date, it is updated before the new sw is activated):

![imagen](https://user-images.githubusercontent.com/6311119/110618161-c4b1db80-8196-11eb-98d7-5701ac05c7f5.png)

and with no user notification:

**Comment by userquin** [comment]
@MartinMalinda I have on local the `injectRegister: 'networkfirst'` option almost ready, as simple as:

```javascript
VitePWA({
  injectRegister: 'networkfirst',
  manifest: {
    ...
  },
})
```

this is what it generates on build (it needs a review, but it works: basically try call `copyWorkboxLibraries` after calling `injectManifest` to remove unnecessary `workbox` libs, see https://github.com/antfu/vite-plugin-pwa/issues/35#issuecomment-797770403):

```javascript
// src/client/build/networkfirst.ts
importScripts("/workbox-v6.1.1/workbox-sw.js");
var debug = JSON.parse("false") === true;
var modulePathPrefix = "/workbox-v6.1.1/";
var useModulePathPrefix = modulePathPrefix.startsWith("/") ? modulePathPrefix.substring(1) : modulePathPrefix;
workbox.setConfig({debug, modulePathPrefix});
var modules = [
  "workbox-core",
  "workbox-routing",
  "workbox-cacheable-response",
  "workbox-strategies",
  "workbox-expiration"
].map((module) => {
  workbox.loadModule(module);
  return `${useModulePathPrefix}${module}`;
});
var cacheNames = workbox.core.cacheNames;
var {registerRoute, setCatchHandler, setDefaultHandler} = workbox.routing;
var {CacheableResponsePlugin} = workbox.cacheableResponse;
var {
  NetworkFirst,
  StaleWhileRevalidate,
  NetworkOnly
} = workbox.strategies;
var {ExpirationPlugin} = workbox.expiration;
var suffix = debug ? ".dev." : ".prod.";
var manifest = [{"revision":"233692e6a59a401d25babb470b42d0ba","url":"assets/[name].11eba18b.js"},{"revision":"0db45f676295b8a5367f17939bfecde5","url":"assets/about.9f61fe0e.js"},{"revision":"701ce0f558c7c79468284db9441d9485","url":"assets/home.ae1a8a75.js"},{"revision":"b32b3c35ce2176e5239f0f5958d3ed08","url":"assets/index.d8d0085a.js"},{"revision":"a760355bef2fd5ded385ccabf84c8de0","url":"assets/index.f3bb671a.css"},{"revision":"cb39939b6c451569e4fa7e86b2413cd8","url":"index.html"},{"revision":"1872c500de691dce40960bb85481de07","url":"registerSW.js"},{"revision":"a7cdb86a9f313bf7d91abf2997c876f8","url":"workbox-v6.1.1/workbox-background-sync.dev.js"},{"revision":"c37bef1035f38894236763f70f847b01","url":"workbox-v6.1.1/workbox-background-sync.prod.js"},{"revision":"b4834c087eaf8da50fcce69a663cbcc0","url":"workbox-v6.1.1/workbox-broadcast-update.dev.js"},{"revision":"9224c426031572af4cbde2db64b7bba4","url":"workbox-v6.1.1/workbox-broadcast-update.prod.js"},{"revision":"bd4298725d2b05c1a7313861ca43119e","url":"workbox-v6.1.1/workbox-cacheable-response.dev.js"},{"revision":"0f99a971609d97b2e235d6f27347cce2","url":"workbox-v6.1.1/workbox-cacheable-response.prod.js"},{"revision":"0ab1809eb6d8f9823adf893131514fc3","url":"workbox-v6.1.1/workbox-core.dev.js"},{"revision":"85c2be1a0e73006ce9e9d1d0cc889459","url":"workbox-v6.1.1/workbox-core.prod.js"},{"revision":"9fb6484c2667d9df08928b0f2f702c76","url":"workbox-v6.1.1/workbox-expiration.dev.js"},{"revision":"c353b8b02a9452019c2ddd9a76620b10","url":"workbox-v6.1.1/workbox-expiration.prod.js"},{"revision":"6f64a834420ad2f28aa8af74c4673bfc","url":"workbox-v6.1.1/workbox-navigation-preload.dev.js"},{"revision":"b455c3e2414561b90c2dac557cf6e87a","url":"workbox-v6.1.1/workbox-navigation-preload.prod.js"},{"revision":"56973179ef1b3124a2f4cc1046b86b75","url":"workbox-v6.1.1/workbox-offline-ga.dev.js"},{"revision":"0b44aa740a7029014d1356317892b96a","url":"workbox-v6.1.1/workbox-offline-ga.prod.js"},{"revision":"6ad698a436eaee2c23637f36102cff02","url":"workbox-v6.1.1/workbox-precaching.dev.js"},{"revision":"b7b24145fd52a89d6127b3310efab41c","url":"workbox-v6.1.1/workbox-precaching.prod.js"},{"revision":"ee1d732eaf6f444875cdecd34df6ed6a","url":"workbox-v6.1.1/workbox-range-requests.dev.js"},{"revision":"591c63c72daadc424e86793e77b92560","url":"workbox-v6.1.1/workbox-range-requests.prod.js"},{"revision":"34f18376df91157ec6ea489e28f945ce","url":"workbox-v6.1.1/workbox-recipes.dev.js"},{"revision":"922bc61480b51855016fb3fc39304398","url":"workbox-v6.1.1/workbox-recipes.prod.js"},{"revision":"fce9cb70987cdba7f37660722ac9f61f","url":"workbox-v6.1.1/workbox-routing.dev.js"},{"revision":"ba807b7a301d7556f34ae12f94b6044e","url":"workbox-v6.1.1/workbox-routing.prod.js"},{"revision":"7385f58adf6fa4ea3ed0de08651d79c8","url":"workbox-v6.1.1/workbox-strategies.dev.js"},{"revision":"ae4e3f5028e8192585bdd0f3d0ef33e5","url":"workbox-v6.1.1/workbox-strategies.prod.js"},{"revision":"8ab7d86a8d8439a2a95614054205a42c","url":"workbox-v6.1.1/workbox-streams.dev.js"},{"revision":"6aca7378ad212a06ce9a64a55dea76ae","url":"workbox-v6.1.1/workbox-streams.prod.js"},{"revision":"55fb6379e95be0790836c1c942f00bd0","url":"workbox-v6.1.1/workbox-sw.js"},{"revision":"534033888a2eb22884af6c83e01c340e","url":"workbox-v6.1.1/workbox-window.dev.umd.js"},{"revision":"140fbae39a000992dca61016d87a6d8f","url":"workbox-v6.1.1/workbox-window.prod.umd.js"}].filter((entry) => {
  return !entry.url || !entry.url.startsWith(useModulePathPrefix) || modules.some((module) => {
    return entry.url.contains(suffix);
  });
});
var cacheName = cacheNames.runtime;
var manifestURLs = [...manifest].map((entry) => {
  const url = new URL(entry.url, self.location);
  return url.href;
});
self.addEventListener("install", (event) => {
  event.waitUntil(caches.open(cacheName).then((cache) => {
    return cache.addAll(manifestURLs);
  }));
});
self.addEventListener("activate", (event) => {
  event.waitUntil(caches.delete(cacheNames.precache).then((result) => {
    if (result)
      console.log("Precached data removed");
    else
      console.log("No precache found");
  }));
});
self.addEventListener("activate", (event) => {
  event.waitUntil(caches.open(cacheName).then((cache) => {
    cache.keys().then((keys) => {
      keys.forEach((request) => {
        if (!manifestURLs.includes(request.url))
          cache.delete(request);
      });
    });
  }));
});
registerRoute(({url}) => manifestURLs.includes(url.href), new NetworkFirst({cacheName}));
setDefaultHandler(new NetworkOnly());
setCatchHandler(async ({event}) => {
  switch (event.request.destination) {
    case "document":
      return await caches.match("/index.html");
    default:
      return Promise.resolve(Response.error());
  }
});
```

**Comment by dheimoz** [comment]
Awesome work!!!!!

**Comment by MartinMalinda** [comment]
@userquin that's awesome 🙇 I'll try to test it in upcoming days

**Comment by bertrand-caron-cascade** [comment]
Hitting this issue now with a Vite/React app (`index.html` gets cached indefinitely). What is the solution in 2025?
